src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetPowerShellUsageCollector.cs

```diff
@@ -0,0 +1,365 @@
+        private void ClearSolutionData()
+        {
+            bool pmcPowershellLoad = _vsSolutionData.LoadedFromPmc;
+            bool pmuiPowershellLoad = _vsSolutionData.LoadedFromPmui;
+
+            _vsSolutionData = new SolutionData();
+
+            _vsSolutionData.LoadedFromPmc = pmcPowershellLoad;
+            _vsSolutionData.LoadedFromPmui = pmuiPowershellLoad;
+        }
+
+        public void Dispose()
+        {
+            NuGetPowerShellUsage.PowerShellLoadEvent -= NuGetPowerShellUsage_PMCLoadEventHandler;
+            NuGetPowerShellUsage.PowerShellCommandExecuteEvent -= NuGetPowerShellUsage_PowerShellCommandExecuteEventHandler;
+            NuGetPowerShellUsage.NuGetCmdletExecutedEvent -= NuGetPowerShellUsage_NuGetCmdletExecutedEventHandler;
+            NuGetPowerShellUsage.InitPs1LoadEvent -= NuGetPowerShellUsage_InitPs1LoadEventHandler;
+            NuGetPowerShellUsage.PmcWindowsEvent -= NuGetPowerShellUsage_PMCWindowsEventHandler;
+            NuGetPowerShellUsage.SolutionOpenEvent -= NuGetPowerShellUsage_SolutionOpenHandler;
+            NuGetPowerShellUsage.SolutionCloseEvent -= NuGetPowerShellUsage_SolutionCloseHandler;
+            NuGetPowerShellUsage.VSInstanceCloseEvent -= NuGetPowerShellUsage_VSInstanseCloseHandler;
+        }
+
+        internal class SolutionData
+        {
+            internal bool FirstTimeLoadedFromPmc { get; set; }
+            internal bool FirstTimeLoadedFromPmui { get; set; }
+            internal bool InitPs1LoadedFromPmcFirst { get; set; }
+            internal bool InitPs1LoadPmc { get; set; }
+            internal bool InitPs1LoadPmui { get; set; }
+            internal bool LoadedFromPmc { get; set; }
+            internal bool LoadedFromPmui { get; set; }
+            internal bool NuGetCommandUsed { get; set; }
+            internal int PmcExecuteCommandCount { get; set; }
+            internal int PmcWindowLoadCount { get; set; }
+            internal int PmuiExecuteCommandCount { get; set; }
+            internal bool SolutionLoaded { get; set; }
+
+            internal SolutionData()
+            {
+                FirstTimeLoadedFromPmc = false;
+                FirstTimeLoadedFromPmui = false;
+                InitPs1LoadedFromPmcFirst = false;
+                InitPs1LoadPmc = false;
+                InitPs1LoadPmui = false;
+                LoadedFromPmc = false;
+                LoadedFromPmui = false;
+                NuGetCommandUsed = false;
+                PmcExecuteCommandCount = 0;
+                PmcWindowLoadCount = 0;
+                PmuiExecuteCommandCount = 0;
+                SolutionLoaded = false;
+
+            }
+
+            internal TelemetryEvent ToTelemetryEvent()
+            {
+                var telemetry = new TelemetryEvent(SolutionClose,
+                    new Dictionary<string, object>()
+                    {
+                        { PowerShellHost + NuGetPowerShellUsageCollector.FirstTimeLoadedFromPmc , FirstTimeLoadedFromPmc },
```

Comment:
Feedback above applies here.

Reply:
@nkolev92 
I created this [PR#3863](https://github.com/NuGet/NuGet.Client/pull/3863) to address this.